source/particles/particle_handler.cc

```diff
@@ -1236,43 +1238,79 @@ namespace Particles
     real_locations.reserve(global_max_particles_per_cell);
     reference_locations.reserve(global_max_particles_per_cell);
 
-    for (const auto &cell : triangulation->active_cell_iterators())
+
+    // We calculate what particles have left their cell in parallel in the
+    // worker. If they have left their cell, we must add them in serial to the
+    // list of missing particles.
+    struct StageOne_CopyData
+    {
+      std::vector<particle_iterator> local_particles_out_of_cell;
+    };
+
+    struct StageOne_ScratchData
+    {
+      std::vector<Point<spacedim>> real_locations;
+      std::vector<Point<dim>>      reference_locations;
+      StageOne_ScratchData(const unsigned int size)
       {
+        real_locations.reserve(size);
+        reference_locations.reserve(size);
+      }
+    };
+
+    const auto stage_one_worker =
+      [&](
+        const typename Triangulation<dim, spacedim>::active_cell_iterator &cell,
+        StageOne_ScratchData &scratch,
+        StageOne_CopyData    &copy) {
         // Particles can be inserted into arbitrary cells, e.g. if their cell is
         // not known. However, for artificial cells we can not evaluate
         // the reference position of particles. Do not sort particles that are
         // not locally owned, because they will be sorted by the process that
         // owns them.
+        scratch.real_locations.clear();
+        copy.local_particles_out_of_cell.clear();
+
         if (cell->is_locally_owned() == false)
           {
-            continue;
+            return;
           }
```

Comment:
You can save a marginal amount of CPU time if you move the clearing of the scratch data to *below* the check whether the cell is locally owned. That's because you don't use the scratch object if the cell is not locally owned, so we don't care whether it is cleared or not -- let the clearing be done on the next cell where we actually have to do something. There's no point to clear the object multiple times if you encounter a bunch of non-locally-owned cells.

You do want to clear the copy object, however, because that is passed on the copier.